### ISSUE TITLE
Pre-delivery gate doesn't validate the artefact as the installer parses it — two install-blocking defects passed it in one delivery

### ISSUE THREAD

**Issue by 14maxwjo** [rebelytics/one-skill-to-rule-them-all#28]
**Labels suggestion:** bug, delivery

---

First: thank you for this skill. The observation-log-plus-scheduled-review loop is genuinely
useful — we backfilled ~7 weeks of work into it and it surfaced eleven concrete corrections to an
existing skill that nobody would have found otherwise. This report is about the last mile only:
the delivery step at the end of the comprehensive review.

## Summary

In one comprehensive review, the loop staged five updated/new skills and delivered them as
`.skill` bundles per `references/weekly-review.md` → "Delivering updated skills". The installer
rejected the delivery **twice, for two unrelated reasons**, and the documented pre-delivery gate
passed on both occasions:

1. `Zip file contains path with invalid characters` — backslash separators in zip member paths,
   because the bundles were packaged on Windows.
2. `field 'description' in SKILL.md must be at most 1024 characters` — three of the five skills
   were over, at 1298, 1224 and 1162 characters.

Neither is an edge case. The first affects **every Windows user** of the review loop whose skill
has a `references/` directory. The second affects anyone following this skill's own guidance to
write long, specific descriptions for reliable triggering.

The shared root cause: the gate inspects the right artefacts, but never tests them against the
**installer's acceptance criteria**. Its two items are

1. every `references/`/`scripts/`/`assets/` path cited in SKILL.md is present in the staged set
2. for multi-file skills, the delivery artefact is the `.skill` bundle rather than bare file links

Both were true both times. The gate establishes *the files are all there* and *the container is
the right kind* — never *the artefact is acceptable to the thing that will install it*.

## Defect 1 — backslash member paths on Windows

`Compress-Archive` (Windows PowerShell 5.1) stores member paths with backslashes:

```
references/classification.md      <- what the spec requires
references\classification.md      <- what actually gets stored
```

ZIP APPNOTE 4.4.17.1 mandates forward slashes as the path separator. A backslash is a **literal
filename character**, so the entry isn't recognised as living in a subdirectory. Confirmed at byte
level — `0x5C` present in both the local file headers and the central directory. Repacking with
forward slashes installs cleanly, contents byte-identical.

Two details that make this harder to fix than it first looks:

**(a) The existing "read the archive listing back after zipping" step surfaced the defect and it
was read straight past.** The listing printed `references\bridge-contract.md`, which on Windows
looks like a perfectly ordinary path. That step's stated purpose in the skill is catching leaked
build artefacts (`__pycache__`, `.DS_Store`) — a wrong-*file* problem — so it was read for a
different failure mode than the one in front of it. Suggestion: name both defect classes on that
step explicitly, since it demonstrably gets read only for the one it mentions.

**(b) The obvious fix, implemented the obvious way, also passes.** A check written with CPython's
`zipfile.namelist()` reports clean forward slashes for the very same malformed archives, because
**CPython normalises `0x5C` to `/` on read**. So "list the members and assert every path uses `/`"
is insufficient unless the reader is documented not to normalise. We only caught this by parsing
the central directory by hand.

**Scope note:** the trigger is **a subdirectory**, not multi-file-ness. Of the five bundles, four
were multi-file but only three contained a `references/` directory — those three were malformed;
the flat multi-file one (`SKILL.md` + `LICENSE`) was fine, as was the single-file one. A gate
keyed on "multi-file" misses that; key it on "contains any path separator".

## Defect 2 — `description` over the 1024-character cap

Three of five skills exceeded the installer's hard limit.

What makes this a *method* failure rather than a simple oversight: **a validation step ran,
measured the description lengths, and reported them as a pass.** The summary read
*"descriptions 900–1,300 chars"*. The number 1,300 was printed, inside a block presented as
verification, above a limit of 1,024 — because the check had **no threshold in it**. It measured
and reported; it never compared.

Compounding it, the value looks *good* to a producer. This skill's guidance (and skill-creator's)
encourages long, specific, "pushy" descriptions to combat undertriggering, so 1,300 characters
reads as best practice. Only the consumer's documented limit disagrees, and nothing in the gate
consults it.

One more data point for why this needs to be machine-checked rather than eyeballed: trimming the
three descriptions took **four iterations** to get all five skills under the cap. Each trim was
estimated, re-measured, and still over — at 1105, 1059, 1041, 1034. Every one of those looks
entirely reasonable to a human reader. With an asserting check it converges in a couple of rounds;
without one it ships.

## Suggested fix — replace the two-item gate with a machine-checked manifest validation

Run as the last step before presenting, asserting at minimum:

- `description` length **≤ 1024 characters**, measured on the **folded** scalar (join the wrapped
  lines) rather than the raw YAML block, since the raw block is longer than what the installer
  sees.
- `name` is kebab-case and **matches the containing directory name**.
- Frontmatter **parses as YAML** and contains both required keys.
- Every `references/`/`scripts/`/`assets/` path cited in the body **exists** in the staged set.
  (Match backticked paths ending in a real file extension — otherwise shell globs that appear in
  prose, like `` `scripts/test-*.mjs` ``, produce false positives. We hit that too.)
- Archive member paths contain **no `0x5C`**, read as **raw stored bytes** rather than through a
  normalising reader.

Two framing suggestions for the skill text itself:

- **State the limits as numbers**, so the check is implementable without guessing, and label them
  as the *consumer's* criteria — they belong in the gate because the installer enforces them, not
  because they seem sensible. Where a limit might drift, say where to confirm it rather than
  hard-coding a guess.
- **Say that the gate must assert, not observe.** Any check that produces a number has to compare
  it to a bound in the same step. An unasserted metric doesn't merely fail to catch defects — it
  manufactures confidence that none exist, which is how 1,300 sailed past 1,024 inside something
  labelled "verification".

## Reference implementation

The two load-bearing parts, both verified against the failing archives:

```python
import zipfile, pathlib, struct, re

MAX_DESCRIPTION_CHARS = 1024        # installer's documented cap

def pack(src: pathlib.Path, out: pathlib.Path):
    """Always writes POSIX separators, on any platform."""
    with zipfile.ZipFile(out, "w", zipfile.ZIP_DEFLATED) as z:
        for f in sorted(p for p in src.rglob("*") if p.is_file()):
            arc = f.relative_to(src).as_posix()
            assert "\\" not in arc, arc
            z.write(f, arcname=arc)

def assert_archive_wellformed(out: pathlib.Path):
    """Central directory as raw bytes. A convenience reader (zipfile.namelist)
    rewrites 0x5C to '/' and would report a malformed archive as clean."""
    data, i = out.read_bytes(), 0
    while True:
        i = data.find(b'PK\x01\x02', i)
        if i < 0:
            return
        n, m, k = (struct.unpack_from('<H', data, i + o)[0] for o in (28, 30, 32))
        name = data[i + 46:i + 46 + n]
        assert b'\x5c' not in name, f"backslash in member: {name!r}"
        i += 46 + n + m + k

def assert_description_within_cap(frontmatter: str):
    m = re.search(r'(?ms)^description:\s*>-?\s*\n(.*?)(?=^\S|\Z)', frontmatter)
    desc = " ".join(m.group(1).split()) if m else ""
    assert len(desc) <= MAX_DESCRIPTION_CHARS, \
        f"description {len(desc)} chars, cap {MAX_DESCRIPTION_CHARS}"
```

Happy to open a PR against the two files that carry the gate
(`references/weekly-review.md` → "Delivering updated skills", and the mirrored copy in
`references/skill-authoring.md` → editing rules item 4) if that's useful — say the word and I'll
send one rather than leaving you to transcribe this.

## Environment

- Windows 11, Windows PowerShell 5.1 (`Compress-Archive`), CPython 3.11
- Five skills delivered; three contained a `references/` subdirectory (all malformed), three
  exceeded the description cap
- Both defects reproduce independently — the description cap has nothing to do with the platform

**Comment by rebelytics** [comment]
Fixed in `release/v3.0.0`. The pre-delivery gate now checks what the installer actually enforces — description cap on the folded value, `name` kebab-case and matching the directory, frontmatter parses, archive member paths use `/` read as raw bytes — and `scripts/validate-skill-bundle.py` asserts all of it and packs a well-formed bundle on any platform. Your reference implementation is the core of that script; it caught a near-boundary description in this very release.

You're credited as co-author on the commit. FYI: I'll be running v3 for a week or so before it goes into main.

**Comment by rebelytics** [comment]
Shipped in v3.0.0, released today. Thanks again for your contribution — the release notes credit everyone whose reports and PRs shaped this version.
